packages/world/src/modules/core/implementations/BatchCallSystem.sol

```diff
@@ -0,0 +1,44 @@
+// SPDX-License-Identifier: MIT
+pragma solidity >=0.8.0;
+
+import { System } from "../../../System.sol";
+import { IBaseWorld } from "../../../codegen/interfaces/IBaseWorld.sol";
+import { revertWithBytes } from "../../../revertWithBytes.sol";
+
+import { SystemCallData, SystemCallFromData } from "../types.sol";
+
+contract BatchCallSystem is System {
+  /**
+   * Batch calls to multiple systems into a single transaction, return the array of return data.
+   */
+  function batchCall(SystemCallData[] calldata systemCalls) public returns (bytes[] memory returnDatas) {
+    IBaseWorld world = IBaseWorld(_world());
+    returnDatas = new bytes[](systemCalls.length);
+
+    for (uint256 i; i < systemCalls.length; i++) {
+      (bool success, bytes memory returnData) = address(world).delegatecall(
+        abi.encodeCall(world.call, (systemCalls[i].systemId, systemCalls[i].callData))
+      );
+      if (!success) revertWithBytes(returnData);
+
+      returnDatas[i] = abi.decode(returnData, (bytes));
+    }
+  }
+
+  /**
+   * Batch calls to multiple systems into a single transaction, return the array of return data.
+   */
+  function batchCallFrom(SystemCallFromData[] calldata systemCalls) public returns (bytes[] memory returnDatas) {
```

Comment:
probably best to go with this approach for extensibility/flexibility, but I wonder how often we'll want to batch call from different addresses vs the signature being

```solidity
function batchCallFrom(address from, SystemCallData[] calldata systemCalls)
```

Reply:
to get that you can do this:

```solidity
world.callFrom(from, coreSystemId, abi.encodeCall(BatchCallSystem.batchCall, (systemCalls)));
```